### PR TITLE
Purge Users Data closes #195

### LINKED ISSUE #195: Purge a user's account, credentials data through the UI (3.1)
As a Lightning user within an organisation that is leaving (or an administrator with a user that is leaving)
I would like to purge my account, credentials and workflow data
So that I can ensure any identifiable or sensitive information will be deleted

- [x] Add a 'purge data' action in the list of users on the `user_live` index page
- [x] Add a 'delete my account' action in the list of users on the `profile_live` page
- [x] For both delete pathways, displays a  confirmation modal (ask them to type out their username) and deletes `users`, `credentials` and `project_users` accordingly. (later, this could be extended to SHOW what will be deleted.)

See last comment for tests.

### Next issue

- [x] get_user_data returns a list of credentials that are owned by the user and projects that the user is in (any other identifiable data ?) to be deleted
- [x] when a user who has their account scheduled for deletion attempts to sign in, they are redirected to the "your account is scheduled for deletion" page

Deletion works as such: 
- project only has one user -> delete the project
- project has several users -> remove user from project users
- credential owned by user and not used anywhere -> delete credential
- credential owned by user but used in a workflow -> ask user to transfer ownership

### PR DESCRIPTION
**Pull request by elias-ba**
This PR implements user account deletion feature. A user can open their settings and mark them self for deletion. This will open a confirmation modal that asks the user to confirm the action by typing down their email. If the email matches the user's email then the user is marked for deletion and their logged out of the app. If the email doesn't match the user's email then an error message is displayed. The user can hit the cancel button to cancel the process. A superuser can also delete other users by going to the users list page and click on the delete action. That action triggers the same behavior explained previously. The only difference is that the superuser is not going to be logged out after each user deletion. After marking a user for deletion, a period of 7 days is observed before the users data (credentials and projects) is completely purged. That is done by Oban.

NB: The PR doesn't implement Emailing. We said that will come later.

Closes #195

### CODE REVIEW

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
@elias-ba , looking good. please address each comment

File: lib/lightning/accounts.ex [hunk 1]
```diff
@@ -18,7 +63,7 @@ defmodule Lightning.Accounts do
 
   """
   def list_users do
-    Repo.all(User)
+    from(u in User, where: is_nil(u.scheduled_deletion)) |> Repo.all()
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning/accounts.ex:66]: @taylordowns2000 we don't wanna display scheduled for deletion users in the users index page.

File: test/lightning/accounts_test.exs [hunk 2]
```diff
@@ -2,12 +2,15 @@ defmodule Lightning.AccountsTest do
...
   alias Lightning.Accounts
+  alias Lightning.Credentials.Credential
+  alias Lightning.Projects.ProjectUser
 
-  import Lightning.AccountsFixtures
   alias Lightning.Accounts.{User, UserToken}
+  import Lightning.{CredentialsFixtures, AccountsFixtures}
 
-  test "list_users/0 returns all users" do
+  test "list_users/0 returns all users not scheduled for deletion" do
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning/accounts_test.exs:11]: @taylordowns2000 if we introduce the notion of scheduling for deletion, then we either:
- don't wanna display users scheduled for deletion
- or have a column in that page that flags users scheduled for deletion

I decided that it's better to hide users scheduled for deletion.

cc: @amberrignell

File: test/lightning/accounts_test.exs [hunk 3]
```diff
@@ -715,10 +797,11 @@ defmodule Lightning.AccountsTest do
     end
   end
 
-  test "delete_user/1 deletes the user" do
+  test "schedule user for deletion" do
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning/accounts_test.exs:800]: @taylordowns2000 We no longer have a `delete_user/1` functionality. We have schedule for deletion.

File: test/lightning_web/controllers/user_session_controller_test.exs [hunk 4]
```diff
@@ -91,6 +98,22 @@ defmodule LightningWeb.UserSessionControllerTest do
       assert response =~ "Log out</span>"
     end
 
+    test "a scheduled for deletion user can't log in", %{
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning_web/controllers/user_session_controller_test.exs:101]: done

File: test/lightning/accounts_test.exs [hunk 5]
```diff
@@ -161,6 +164,85 @@ defmodule Lightning.AccountsTest do
...
+      assert remaining_projs
+             |> Enum.any?(fn x -> x.user_id == user_2.id end)
+
+      assert remaining_users
+             |> Enum.any?(fn x -> x.id == user_2.id end)
+    end
+  end
+
+  describe "Oban" do
+    test "Oban removes all users past deletion date" do
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning/accounts_test.exs:224]: Done

File: config/runtime.exs [hunk 6]
```diff
@@ -31,7 +31,8 @@ config :lightning, Oban,
     {Oban.Plugins.Cron,
      crontab: [
        {"* * * * *", Lightning.Jobs.Scheduler},
-       {"* * * * *", ObanPruner}
+       {"* * * * *", ObanPruner},
+       {"* * * * *", Lightning.Accounts, args: %{"type" => "purge_deleted"}}
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply config/runtime.exs:35]: Done

File: lib/lightning/accounts.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (taylordowns2000) [reply lib/lightning/accounts.ex:66]: @elias-ba , please restore this function to how it was before. your test will be failing, and we'll make it pass via a new column in that superuser users list view.

File: test/lightning/accounts_test.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning/accounts_test.exs:800]: Action here: restore `delete_user/1` and it's corresponding tests

File: lib/lightning/accounts.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning/accounts.ex:75]: Done

File: test/lightning/accounts_test.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning/accounts_test.exs:799]: Done

File: test/lightning_web/live/profile_live_test.exs [hunk 7]
```diff
@@ -92,4 +96,79 @@ defmodule LightningWeb.ProfileLiveTest do
...
+      assert html =~ "Your account and credential data will be deleted"
+
+      assert new_live
+             |> form("#scheduled_deletion_form",
+               user: @invalid_schedule_deletion_attrs
+             )
+             |> render_change() =~
+               "This email doesn&#39;t match your current email"
+
+      {:error,
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply test/lightning_web/live/profile_live_test.exs]: Done

**Comment by elias-ba** [comment]
> @elias-ba , looking good. please address each comment

@taylordowns2000 this is now good and ready for another round of review. You'll probably need to handle the different #TODOs, they're blocking Circle CI.

**Comment by elias-ba** [comment]
@taylordowns2000 I have rebased against main to handle conflicts but I didn't squash commits. There are a lof of them and I really think most of them are good chunks of work. Maybe you'll take time to squash some of them but to my mind this is really not necessary.

**Comment by taylordowns2000** [comment]
@amberrignell , this is ready for your review. if you're happy with how it behaves (and i think you've already checked most of it out) then i'll squash and merge. note that there are some code style issues that i've moved into https://github.com/OpenFn/Lightning/issues/290 but we can address them as a team after this is in main if these tests give you what you need to ship the deliverable

**Pull request merged**